### PR TITLE
[Feature:InstructorUI] Download Button for Rainbow Grades

### PR DESCRIPTION
**Pull request by TusharPuri10**
fixes https://github.com/Submitty/Submitty/issues/10078
closes #10213 

### What is the current behavior?
=> Currently the .csv of the grades from rainbow grades is located in /var/local/submitty/courses/SEMESTER/COURSE/rainbow_grades/output.csv

### What is the new behavior?
=> Added a blue "Download CSV of grades" button to get this file from the top of the "Gradebook" page.
image

### Other information?
There was PR with an minor update of RainbowGrades.twig made from PR #10213, which has been cherry picked to this PR, as well as the video demo. The work should be credited accordingly to @MUFFANUJ 

https://github.com/Submitty/Submitty/assets/151390316/06a29734-d22b-4074-9ffc-853bfa60783f

Some useful feedbacks from #10213 was not carried over, please consider reviewing those comments as well when taking over this PR.

### CODE REVIEW

File: site/app/controllers/admin/ReportController.php [hunk 1]
```diff
@@ -789,4 +789,49 @@ public function displayGradebook() {
...
+        $csvFilePath = '/var/local/submitty/courses/f23/development/rainbow_grades/output.csv';
+
+        // Check if the file exists
+        if (file_exists($csvFilePath)) {
+            // Set appropriate headers for file download with a custom filename
+            header('Content-Description: File Transfer');
+            header('Content-Type: application/csv');
+            header('Content-Disposition: attachment; filename="' . $this->generateCustomFilename() . '"');
+            header('Expires: 0');
+            header('Cache-Control: must-revalidate');
```
Comment (cjreed121) [inline site/app/controllers/admin/ReportController.php:825]: This route should not be necessary. See the `downloadCourseFile` function in `site/app/controllers/MiscController.php`. Our `Access.php` already has rainbow_grades as a valid directory so you should be able to create a URL to the `/courses/{semester}/{course}/download` path.

File: site/app/templates/admin/Report.twig [hunk 2]
```diff
@@ -10,6 +10,23 @@
...
+            <div class="card-btn-cont">
+                <button onclick="location.href='{{ rainbow_grades_csv_url }}'"
+                class="btn btn-primary key_to_click"
+                tabindex="0"
+                >
+                    Download CSV of grades
+                </button>
+                <div class="card-btn-info">
+                    Last ran: <span id="grade-summaries-last-run">{{ grade_summaries_last_run }}</span>
+                </div>
```
Comment (cjreed121) [inline site/app/templates/admin/Report.twig:26]: You might even be able to take advantage of the `downloadFile` function from server.js instead which would let you download the file even easier.

**Review by cjreed121 — CHANGES_REQUESTED** [review]
Thanks for the contribution! Your changes seem that it would address the issue but there is a more preferred way to address the problem which I noted below.

**Comment by TusharPuri10** [comment]
Hi chris,

Thanks for your input. I'll update the code as suggested by using the
existing `downloadCourseFile` function. I'll also explore integrating the
`downloadFile` function from `server.js` for a smoother download process.

I'll keep you posted on the progress.

Regards

File: site/app/templates/admin/Report.twig [hunk 3]
```diff
@@ -3,13 +3,37 @@
         $(document).ready(function() {
             // Bind click listener to grade summaries button
             $('#grade-summaries-button').click(function() {
-                $('#grade-summaries-last-run').html('Running...');
+                $('.grade-summaries-last-run').html('Running...');
+            });
+            // Bind click listener to download buttons
+            $('.download-btn').click(function() {
+                let path = $(this).data('path');
+                let dir = $(this).data('dir');
```
Comment (cjreed121) [inline site/app/templates/admin/Report.twig:11]: `const` is preferred over `let`

File: site/app/templates/admin/Report.twig [hunk 4]
```diff
@@ -23,7 +47,7 @@
                     Generate Grade Summaries
                 </button>
                 <div class="card-btn-info">
-                    Last ran: <span id="grade-summaries-last-run">{{ grade_summaries_last_run }}</span>
+                    Last ran: <span class="grade-summaries-last-run">{{ grade_summaries_last_run }}</span>
```
Comment (cjreed121) [inline site/app/templates/admin/Report.twig:50]: Why is this changing?

File: site/app/templates/admin/Report.twig [hunk 5]
```diff
@@ -3,13 +3,37 @@
...
     </script>
 
     <div id="grade-report">
         <h1>Grade Reports</h1>
+        <div id="rainbow-grade-report-csv" class="card">
+            <div class="title-cont">
+                Pushing this button will generate a CSV file (for download) of Rainbow Grades.
+            </div>
+            <div class="card-btn-cont">
+                <button class="btn btn-primary key_to_click download-btn"
```
Comment (cjreed121) [inline site/app/templates/admin/Report.twig:24]: An ID would be a better choice for the download button since there will only be one

File: site/app/templates/admin/Report.twig [hunk 6]
```diff
@@ -3,13 +3,37 @@
...
+                        data-path="/var/local/submitty/courses/f23/development/rainbow_grades/output.csv"
+                        data-dir="rainbow_grades"
```
Comment (cjreed121) [inline site/app/templates/admin/Report.twig:26]: These are hardcoded values. path should be dynamically generated for the specific semester/course context we are in. Also dir is completely static so that could be moved entirely to the JS.

**Review by cjreed121 — CHANGES_REQUESTED** [review]

File: site/app/templates/admin/Report.twig (hunk 6 above)
Comment (TusharPuri10) [reply site/app/templates/admin/Report.twig:26]: I apologize for the oversight

File: site/app/templates/admin/Report.twig (hunk 3 above)
Comment (TusharPuri10) [reply site/app/templates/admin/Report.twig:11]: Got it

**Comment by TusharPuri10** [comment]
@cjreed121 I solved the issue but the user first has to **Generate the Rainbow Grades** to actually create the path to output.csv else it will throw
`$this->core->getOutput()->showError(self::GENERIC_NO_ACCESS_MSG);`
image

because the file does not exist

**path:  \Submitty\site\app\libraries\Access.php**
```php
public function canUser($user, string $action, $args) {
...
            //This is not a valid directory
            //checks for top level dirs like annotations, checkout, course_materials, submissions ... )
            if (!array_key_exists($dir, $this->directories)) {
                return false;
            }
            elseif (!file_exists($path)) {
                //checks for the existence of path which is asked for
                return false;
            }
...
}
```
How do you think this can be solved? Should I add message "Please generate rainbow grades first" or add default columns to table to generate it on click itself.

**Comment by TusharPuri10** [comment]
https://github.com/Submitty/Submitty/issues/9805 is part of this issue. I think solving it and then adding the similar solution to the "Download CSV of grades" button will solve the issue I stated in above comment.

**Comment by bmcutler** [comment]
@TusharPuri10 

In the original issue I requested this be placed on the "Gradebook" page, not on the "Grade Reports" page.
The gradebook page link appears on the left sidebar when "Display Rainbow Grades" is selected on "Course settings"

Thanks.

File: site/app/libraries/Access.php [hunk 7]
```diff
@@ -914,8 +914,17 @@ public function resolveDirPath(string $dir, string $path) {
...
-        if (!str_starts_with($path, $info["base"])) {
+        // Diffferent base path for downloading output.csv of rainbow grades
+        if ($dir === "rainbow_grades" && substr($path, -11) === "/output.csv") {
+            // Removing "/upload" from path
+            $base = substr($info["base"],0,-23) . "/rainbow_grades";
+            //checking the path
+            if(!str_starts_with($path, $base))
+            {
+                $path = $base . $path;
+            }
```
Comment (cjreed121) [inline site/app/libraries/Access.php:925]: Is this change just to remove the `/upload` part from the path? You should be able to just add `/upload` in the JS call.

**Review by cjreed121 — CHANGES_REQUESTED** [review]
Making the change that @bmcutler noted should address the issue you are having.

File: site/app/libraries/Access.php (hunk 7 above)
Comment (TusharPuri10) [reply site/app/libraries/Access.php:925]: yes, it is to remove the `/upload` part from path. Otherwise I had to change the base path of `rainbow_grades` directory
``` php
private function loadDirectories() {
...
        $this->directories["rainbow_grades"] = [
            "base" => $this->core->getConfig()->getCoursePath() . "/uploads/rainbow_grades",
            "subparts" => [],
            "permissions" => [
                "path.read" => "path.read.rainbow_grades",
                "path.write" => "path.write.rainbow_grades",
            ]
        ];
...
}
```
to bypass the checks of `Access.php`. This will effect other URLs as well that's why I thought this will be a faster solution. If you insist I can make those changes by changing the base path.
> @TusharPuri10
> 
> In the original issue I requested this be placed on the "Gradebook" page, not on the "Grade Reports" page. The gradebook page link appears on the left sidebar when "Display Rainbow Grades" is selected on "Course settings"
> 
> Thanks.

 I will also address this issue.

**Comment by TusharPuri10** [comment]
Added a blue "Download CSV of grades" button to get this file from the top of the "Gradebook" page. I apologize for the inconvenience as I have misread the issue first time. @cjreed121 please review.
image

**Comment by TusharPuri10** [comment]
@cjreed121 Can you please tell me about Cypress (UI) check. I was unable to solve it.

**Review by TusharPuri10 — COMMENTED** [review]
Can you please tell me about Cypress (UI) check. @cjreed121

**Review by TusharPuri10 — COMMENTED** [review]
I apologize for that. It won't happen again.

File: site/app/libraries/Access.php [hunk 8]
```diff
@@ -914,8 +914,16 @@ public function resolveDirPath(string $dir, string $path) {
...
+        // Diffferent base path for downloading output.csv of rainbow grades
+        if ($dir === "rainbow_grades" && substr($path, -11) === "/output.csv") {
+            // Removing "/upload" from path
+            $base = substr($info["base"], 0, -23) . "/rainbow_grades";
+            //checking the path
+            if (!str_starts_with($path, $base)) {
+                $path = $base . $path;
+            }
+        }//Make sure it starts with the dir base
+        elseif (!str_starts_with($path, $info["base"])) {
```
Comment (cjreed121) [inline site/app/libraries/Access.php:926]: Sorry, I'm actually going to go back on what I said. I thought the Access.php rainbow_grades directory pointed to this one but didn't realize it had the upload on it. I don't think we should add something like this to Access.php. If you could just grab your code from before in the git log that had the new route which downloaded the file, that would be great.

**Review by cjreed121 — CHANGES_REQUESTED** [review]

**Comment by ziesski** [comment]
There was PR with an minor update of RainbowGrades.twig made from PR #10213, which has been cherry picked to this PR. The work should be credited accordingly to @MUFFANUJ 

Some useful feedbacks from #10213 was not carried over, please consider reviewing those comments as well when taking over this PR.

File: site/app/libraries/Access.php (hunk 8 above)
Comment (dorian451) [reply site/app/libraries/Access.php]: Fixed in https://github.com/Submitty/Submitty/pull/10079/commits/7fcf166721524d006e8da11a8feb43b52365c2fb

**Comment by zacharymnp** [comment]
the new button produces a different .csv than the old button - although it is still a Rainbow Grades .csv

**Review by zacharymnp — APPROVED** [review]

File: site/app/controllers/admin/ReportController.php [hunk 9]
```diff
@@ -135,13 +135,10 @@ public function getGradeSummariesLastRun() {
 
     /**
      * Generates and offers download of CSV grade report
+     * @AccessControl(role="INSTRUCTOR")
```
[the remark this answers is not in the document]
Comment (dorian451) [reply site/app/controllers/admin/ReportController.php]: Fixed in 964525f73b3d7ccc480cfed597b0003e316abca4

File: site/app/controllers/admin/ReportController.php [hunk 10]
```diff
@@ -776,8 +772,56 @@ public function displayGradebook() {
...
             )
         );
     }
+
+    /**
+     * Generate a custom filename for the downloaded CSV file
+     *
+     * @return string
+     */
+    private function generateCustomFilename() {
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:786]: It would be better to use native PHP type annotations instead of PHPDoc comments, because native type annotations can be enforced by the language itself.
```suggestion
    /**
     * Generate a custom filename for the downloaded CSV file
     */
    private function generateCustomFilename(): string {
```

File: site/app/controllers/admin/ReportController.php [hunk 11]
```diff
@@ -776,8 +772,56 @@ public function displayGradebook() {
...
+        $course = $this->core->getConfig()->getCourse();
+        $timestamp = date("ymdHis");
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:788]: You could use `DateUtils::getFileNameTimeStamp()` for this instead.

File: site/app/controllers/admin/ReportController.php [hunk 12]
```diff
@@ -776,8 +772,56 @@ public function displayGradebook() {
...
+        return "{$course}_rainbow_grades_{$timestamp}.csv";
+    }
+
+
+    /**
+     * Download CSV file for Rainbow Grades
+     * @return void
+     */
+    #[Route("/courses/{_semester}/{_course}/reports/rainbow_grades_csv")]
+    public function downloadRainbowGradesCSVFile() {
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:798]: Same comment here:
```suggestion
    /**
     * Download CSV file for Rainbow Grades
     */
    #[Route("/courses/{_semester}/{_course}/reports/rainbow_grades_csv")]
    public function downloadRainbowGradesCSVFile(): void {
```

File: site/app/controllers/admin/ReportController.php [hunk 13]
```diff
@@ -776,8 +772,56 @@ public function displayGradebook() {
...
+        // Check if the file exists
+        if (file_exists($csvFilePath)) {
+            // Set appropriate headers for file download with a custom filename
+            header('Content-Description: File Transfer');
+            header('Content-Type: application/csv');
+            header('Content-Disposition: attachment; filename="' . $this->generateCustomFilename() . '"');
+            header('Expires: 0');
+            header('Cache-Control: must-revalidate');
+            header('Pragma: public');
+            header('Content-Length: ' . filesize($csvFilePath));
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:816]: [this](https://github.com/Submitty/Submitty/pull/10079#discussion_r1438422010) review comment from Chris still needs to be addressed.

File: site/app/controllers/admin/ReportController.php [hunk 14]
```diff
@@ -776,8 +772,56 @@ public function displayGradebook() {
...
+
+            // Output the file content
+            readfile($csvFilePath);
+            exit;
+        }
+        else {
+            // Handle the case where the file does not exist
+            $this->core->getOutput()->showError($csvFilePath . " was not found or was not readable.\nMaybe you have not <a\thref='https://submitty.org/instructor/course_settings/rainbow_grades/automatic_setup'>generated the rainbow grades</a> yet?");
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:824]: Putting a link in an error message banner seems like a bad idea...

File: site/app/templates/submission/RainbowGrades.twig [hunk 15]
```diff
@@ -1,11 +1,32 @@
...
+                        onclick='window.open("{{ rainbow_grades_csv_url }}", "_blank").focus()'
+                        tabindex="0"
+                    >
+                        Download CSV of grades
+                    </button>
+                </div>
+                <p>
+                    Grades last updated on: 
+                    <b>
+                    <span>{{ grade_summaries_last_run }}</span>
```
Comment (williamjallen) [inline site/app/templates/submission/RainbowGrades.twig:24]: Is this `<span>` necessary?

File: site/public/css/rainbow-grades.css [hunk 16]
```diff
@@ -1,3 +1,11 @@
...
     color: black;
 }
+
+#rainbow-grades-header > .btn-wrapper {
+    margin: auto;
+}
+
+#rainbow-grades-header > p {
+    float: right;
+}
```
Comment (williamjallen) [inline site/public/css/rainbow-grades.css:11]: Targeting elements by raw selectors like `p` is discouraged.  Please use a class or ID instead.  You can also probably find an existing CSS selector to do this.

File: site/app/templates/submission/RainbowGrades.twig [hunk 17]
```diff
@@ -1,11 +1,32 @@
...
+    <header>
+        <h1 class="label">
+            {% if extra_label is defined %}
+                Rainbow Grades {{extra_label}}
+            {% else %}
+                Rainbow Grades
+            {% endif %}
+        </h1>
+        <div id="rainbow-grades-header">
+            {% if show_summary and extra_label is defined and extra_label == "For All Students" %}
```
Comment (williamjallen) [inline site/app/templates/submission/RainbowGrades.twig:11]: This check against a string to be shown in the UI feels like a hack.  Why not create a variable specifically for the "all students" case?

File: site/app/controllers/admin/ReportController.php [hunk 18]
```diff
@@ -776,8 +772,56 @@ public function displayGradebook() {
...
+    /**
+     * Download CSV file for Rainbow Grades
+     * @return void
+     */
+    #[Route("/courses/{_semester}/{_course}/reports/rainbow_grades_csv")]
+    public function downloadRainbowGradesCSVFile() {
+        // Check if the user has the necessary permissions (e.g., admin access)
+        if (!$this->core->getUser()->accessAdmin()) {
+            $this->core->getOutput()->showError("This account cannot access admin pages");
+        }
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:802]: I believe this isn't necessary since the entire class is annotated with `@AccessControl(role="INSTRUCTOR")`.

File: site/app/controllers/admin/ReportController.php [hunk 19]
```diff
@@ -138,10 +138,6 @@ public function getGradeSummariesLastRun() {
      */
     #[Route("/courses/{_semester}/{_course}/reports/csv")]
     public function generateCSVReport() {
-        if (!$this->core->getUser()->accessAdmin()) {
-            $this->core->getOutput()->showError("This account cannot access admin pages");
-        }
-
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:144]: While this is a good change to make, it doesn't seem to be related to the rest of this PR, and should be done separately.

**Review by williamjallen — CHANGES_REQUESTED** [review]

File: site/app/controllers/admin/ReportController.php (hunk 13 above)
Comment (dorian451) [reply site/app/controllers/admin/ReportController.php:813]: https://github.com/Submitty/Submitty/pull/10079#discussion_r1530954806

File: site/app/controllers/admin/ReportController.php (hunk 14 above)
Comment (dorian451) [reply site/app/controllers/admin/ReportController.php:821]: Why?

File: site/app/controllers/admin/ReportController.php [hunk 20]
```diff
@@ -776,8 +775,49 @@ public function displayGradebook() {
...
+
+
+    /**
+     * Download CSV file for Rainbow Grades
+     */
+    #[Route("/courses/{_semester}/{_course}/reports/rainbow_grades_csv")]
+    public function downloadRainbowGradesCSVFile(): void {
+
+        // Path to the CSV file for Rainbow Grades
+        $csvFilePath = '/var/local/submitty/courses/' . $this->core->getConfig()->getTerm() . '/' . $this->core->getConfig()->getCourse() . '/rainbow_grades/output.csv';
```
Comment (ziesski) [inline site/app/controllers/admin/ReportController.php:801]:
```suggestion
        $csvFilePath = FileUtils::joinPaths(
    '/var/local/submitty/courses',
    $this->core->getConfig()->getTerm(),
    $this->core->getConfig()->getCourse(),
    'rainbow_grades',
    'output.csv'
);

```

**Review by ziesski — CHANGES_REQUESTED** [review]

File: site/app/controllers/admin/ReportController.php (hunk 20 above)
Comment (ziesski) [reply site/app/controllers/admin/ReportController.php:801]: Using `FileUtils::joinPaths` ensures that the correct path separators are used, regardless of the operating system.

Comment (dorian451) [reply site/app/controllers/admin/ReportController.php:801]: I forgot, my bad

File: site/app/controllers/admin/ReportController.php (hunk 14 above)
Comment (williamjallen) [reply site/app/controllers/admin/ReportController.php:826]: If the banner message disappears after a period of time, so too will the documentation the users wants to click on.  Couldn't you display this else in the UI somewhere instead?  At a minimum, the HTML should be injected via Twig.

File: site/app/controllers/admin/ReportController.php (hunk 13 above)
Comment (williamjallen) [reply site/app/controllers/admin/ReportController.php:818]: Don't we already have some existing code for file downloads?

File: site/app/controllers/admin/ReportController.php [hunk 21]
```diff
@@ -776,8 +775,55 @@ public function displayGradebook() {
...
+        // Check if the file exists
+        if (file_exists($csvFilePath)) {
+            // Set appropriate headers for file download with a custom filename
+            header('Content-Description: File Transfer');
+            header('Content-Type: application/csv');
+            header('Content-Disposition: attachment; filename="' . $this->generateCustomFilename() . '"');
+            header('Expires: 0');
+            header('Cache-Control: must-revalidate');
+            header('Pragma: public');
+            header('Content-Length: ' . filesize($csvFilePath));
```
Comment (IDzyre) [inline site/app/controllers/admin/ReportController.php:818]: Could this use the DownloadResponse from 'app\libraries\response\DownloadResponse' instead of setting headers manually?

**Review by IDzyre — CHANGES_REQUESTED** [review]

File: site/app/controllers/admin/ReportController.php (hunk 21 above)
Comment (IDzyre) [reply site/app/controllers/admin/ReportController.php:818]: If you to file_get_contents($csvFilePath) instead of readfile() does it work?

File: site/app/controllers/admin/ReportController.php (hunk 14 above)
Comment (dorian451) [reply site/app/controllers/admin/ReportController.php:826]: The error message will never disappear, as it is being rendered by ErrorPage.twig, which never hides the error message. I believe that the only way to move the link out of file would be to either to change how error pages are created in Submitty or make a brand-new error page specifically for this one error. I could also just remove the link, but I figured having it there would make it easier for users to debug their problem.

File: site/app/controllers/admin/ReportController.php (hunk 21 above)
Comment (dorian451) [reply site/app/controllers/admin/ReportController.php:818]: Unfortunately no, it just returns an empty page.

Comment (IDzyre) [reply site/app/controllers/admin/ReportController.php:818]: You have to return the download response for it to work.

Comment (dorian451) [reply site/app/controllers/admin/ReportController.php:818]: If I do that PHP gets mad and says `Fatal error: A void function must not return a value in /usr/local/submitty/site/app/controllers/admin/ReportController.php on line 826`

Comment (dorian451) [reply site/app/controllers/admin/ReportController.php]: Fixed in 23c6f701bcd9d1650355e79cdd1006c539af06bc

**Review by ziesski — APPROVED** [review]
Code looks good, we just need to decide when to clean up the old generate csv report button from report page. I made the issue and linked to this PR #10591 
Sudden change may be confuse existing users (in this case I don't think there are many users using this feature) so might be nice if we can have notification & redirect link next to the old button saying "this button will be moved to [here](http://localhost:1511/courses/s24/sample/gradebook)"
